File: app/src/main/java/com/capstone/project_niyakneyak/data/patient_resource/PatientRepository.java
package com.capstone.project_niyakneyak.data.patient_resource;


import androidx.lifecycle.MutableLiveData;

import com.capstone.project_niyakneyak.R;
import com.capstone.project_niyakneyak.data.Result;
import com.capstone.project_niyakneyak.data.patient_model.MedsData;
import com.capstone.project_niyakneyak.data.patient_model.PatientData;
import com.capstone.project_niyakneyak.ui.main.etc.ActionResult;
import com.capstone.project_niyakneyak.ui.main.etc.DataView;

import java.io.Serializable;

import kotlinx.coroutines.sync.Semaphore;

public class PatientRepository implements Serializable {
    private static volatile PatientRepository instance;
    private final PatientDataSource dataSource;
    private PatientData patientData = null;
    private final MutableLiveData<ActionResult> actionResult = new MutableLiveData<>();

    private PatientRepository(PatientDataSource dataSource){
        this.dataSource = dataSource;
        if(patientData == null){
            Result<PatientData> result_read = dataSource.read_PatientData();
            if(result_read instanceof Result.Success)
                patientData = ((Result.Success<PatientData>) result_read).getData();
            else patientData = new PatientData();
        }
    }

    public static PatientRepository getInstance(PatientDataSource dataSource){
        if (instance == null) {
            instance = new PatientRepository(dataSource);
        }
        return instance;
    }


    //Getter for patient Info.
    public Result<PatientData> getPatientData(){
        Result<PatientData> result = dataSource.read_PatientData();
        if(result instanceof Result.Success) return result;
        else{
            if(patientData != null) {return new Result.Success<>(patientData);}
            return new Result.Error(new Exception("Cannot get data from remote and repository"));
        }
    }


    //Useful Functions to search, add, modify, delete patient medication Info.
    public Result<MedsData> search_MedsData(MedsData data){
        Result<MedsData> result = dataSource.search_MedsData(data);
        if(result instanceof Result.Success){
            actionResult.setValue(new ActionResult(new DataView("Searching MedsData Success")));
            return result;
        }
        else{
            if(patientData.searchMedsData(data)) {
                actionResult.setValue(new ActionResult(new DataView("Searching MedsData Success")));
                return new Result.Success<>(data);
            }
            actionResult.setValue(new ActionResult(R.string.action_main_rcv_search_error));
            return new Result.Error(new Exception("Data does not exists!"));
        }
    }
    public Result<MedsData> add_MedsData(MedsData data) {
        Result<MedsData> result = dataSource.write_MedsData(data, 0);
        patientData.addMedsData(data);
        if(result instanceof Result.Success){
            actionResult.setValue(new ActionResult(new DataView("Adding MedsData Success")));
            return result;
        }
        else {
            if(patientData != null) {
                actionResult.setValue(new ActionResult(new DataView("Adding MedsData Success")));
                return new Result.Success<>(data);
            }
            actionResult.setValue(new ActionResult(R.string.action_main_rcv_add_error));
            return new Result.Error(new Exception("Cannot connect to database but saved in local repository"));
        }
    }

    public Result<MedsData> modify_MedsData(MedsData originData, MedsData changedData){
        Result<MedsData> result = dataSource.write_MedsData(originData, changedData);
        boolean flag = patientData.modifyMedsData(originData, changedData);

        if(result instanceof Result.Success){
            actionResult.setValue(new ActionResult(new DataView("Modifying MedsData Success")));
            return result;
        }
        else {
            if(flag){
                actionResult.setValue(new ActionResult(new DataView("Modifying MedsData Success")));
                return new Result.Success<>(changedData);
            }
            actionResult.setValue(new ActionResult(R.string.action_main_rcv_modify_error));
            return new Result.Error(new Exception("Data does not exists!"));
        }
    }

    public Result<MedsData> delete_MedsData(MedsData target){
        Result<MedsData> result = dataSource.write_MedsData(target, 1);
        boolean flag = patientData.deleteMedsData(target);

        if(result instanceof Result.Success) {
            actionResult.setValue(new ActionResult(new DataView("Deleting MedsData Success")));
            return result;
        }
        else return flag ? new Result.Success<>(target) : new Result.Error(new Exception("Data does not exists!"));
    }

    public MutableLiveData<ActionResult> getActionResult(){return actionResult;}
}

File: app/src/main/java/com/capstone/project_niyakneyak/data/patient_resource/PatientRepository.kt
package com.capstone.project_niyakneyak.data.patient_resource

import androidx.lifecycle.MutableLiveData
import com.capstone.project_niyakneyak.R
import com.capstone.project_niyakneyak.data.Result
import com.capstone.project_niyakneyak.data.patient_model.MedsData
import com.capstone.project_niyakneyak.data.patient_model.PatientData
import com.capstone.project_niyakneyak.ui.main.etc.ActionResult
import com.capstone.project_niyakneyak.ui.main.etc.DataView
import java.io.Serializable

class PatientRepository private constructor(private val dataSource: PatientDataSource) :
    Serializable {
    private var patientData: PatientData? = null
    val actionResult = MutableLiveData<ActionResult>()

    init {
        if (patientData == null) {
            val resultRead = dataSource.readPatientData()
            patientData =
                if (resultRead is Result.Success<*>) (resultRead as Result.Success<PatientData?>).data else PatientData()
        }
    }

    //Getter for patient Info.
    fun getPatientData(): Result<Any?> {
        val result = dataSource.readPatientData()
        return if (result is Result.Success<*>) result else {
            if (patientData != null) {
                Result.Success(patientData)
            } else Result.Error(Exception("Cannot get data from remote and repository"))
        }
    }

    //Useful Functions to search, add, modify, delete patient medication Info.
    fun searchMedsData(data: MedsData): Result<Any?> {
        val result = dataSource.searchMedsData(data)
        return if (result is Result.Success<*>) {
            actionResult.value = ActionResult(DataView("Searching MedsData Success"))
            result
        } else {
            if (patientData!!.searchMedsData(data)) {
                actionResult.value = ActionResult(DataView("Searching MedsData Success"))
                return Result.Success(data)
            }
            actionResult.value = ActionResult(R.string.action_main_rcv_search_error)
            Result.Error(Exception("Data does not exists!"))
        }
    }

    fun addMedsData(data: MedsData): Result<Any?> {
        val result = dataSource.writeMedsData(data, 0)
        patientData!!.addMedsData(data)
        return if (result is Result.Success<*>) {
            actionResult.value = ActionResult(DataView("Adding MedsData Success"))
            result
        } else {
            if (patientData != null) {
                actionResult.value = ActionResult(DataView("Adding MedsData Success"))
                return Result.Success(data)
            }
            actionResult.value = ActionResult(R.string.action_main_rcv_add_error)
            Result.Error(Exception("Cannot connect to database but saved in local repository"))
        }
    }

    fun modifyMedsData(originData: MedsData, changedData: MedsData): Result<Any?> {
        val result = dataSource.writeMedsData(originData, changedData)
        val flag = patientData!!.modifyMedsData(originData, changedData)
        return if (result is Result.Success<*>) {
            actionResult.value = ActionResult(DataView("Modifying MedsData Success"))
            result
        } else {
            if (flag) {
                actionResult.value = ActionResult(DataView("Modifying MedsData Success"))
                return Result.Success(changedData)
            }
            actionResult.value = ActionResult(R.string.action_main_rcv_modify_error)
            Result.Error(Exception("Data does not exists!"))
        }
    }

    fun deleteMedsData(target: MedsData): Result<Any?> {
        val result = dataSource.writeMedsData(target, 1)
        val flag = patientData!!.deleteMedsData(target)
        return if (result is Result.Success<*>) {
            actionResult.value = ActionResult(DataView("Deleting MedsData Success"))
            result
        } else if (flag) Result.Success(target) else Result.Error(
            Exception("Data does not exists!")
        )
    }

    companion object {
        @Volatile
        private var instance: PatientRepository? = null
        fun getInstance(dataSource: PatientDataSource): PatientRepository? {
            if (instance == null) {
                instance = PatientRepository(dataSource)
            }
            return instance
        }
    }
}